### PR TITLE
Fix kubeadm file discovery

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
kubernetes/kubernetes#66482 enabled  `kubeadm join --discovery-file` with a discovery file with embedded credentials. However, the provided credentials are used only to get the cluster-info config map, then kubeadm rollbacks to the anonymous user and the workflow exit with an obscure message `"kubeadm-config" is forbidden: User "system:anonymous" cannot get resource "configmaps" in API group "" in the namespace "kube-system"`

This PR fixes this problem and makes possible to do `kubeadm join --discovery-file` with a discovery file with embedded credentials; it also implements a better logging and error messages in cases of misconfigurations

**Which issue(s) this PR fixes**:
Fixes #https://github.com/kubernetes/kubeadm/issues/1687

**Special notes for your reviewer**:
IMO we should consider this for backporting because the error message is really confusing

**Does this PR introduce a user-facing change?**:
```release-note
Fix error in `kubeadm join --discovery-file` when using discovery files with embedded credentials
```
/sig cluster-lifecycle
/priority critical-urgent
/assign @timothysc 
/assign @neolit123 

/cc @hickeng

### CODE REVIEW

**Comment by neolit123** [comment]
> IMO we should consider this for backporting because the error message is really confusing

+1 for backport to 1.14 and 1.15.

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go]: `no AuthInfo object`

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go]: `AuthInfo object with bad contents`, or similar.

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go]: `valid AuthInfo`

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go]: `no such CurrentContext object`

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go]: `Context object with bad contents`, or similar.

File: cmd/kubeadm/app/discovery/discovery.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/discovery.go]: `and the TLSBootstrapToken flag` -> `but with the TLSBootstrapToken flag`?

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go]: suggested minor improvements for the test names here.
please adjust as see fit.

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: `info are embedded in config file`->`info is embedded in the kubeconfig file`

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: `a authentication` -> `authentication`

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: can the logic here reach a case where `HasAuthenticationCredentials()` returns `true` for a token in the kubeconfig, but then we want to load the certs on disk into the file?

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: `currentCluster := kubeconfigutil.GetClusterFromKubeConfig(kubeconfig)`

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: currently this logic runs indefinitely and we should change this to fail after a while.
i'd say 10 retries * 5 seconds (DiscoveryRetryInterval) should be enough.

a TODO can lead to someone picking this up.

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: `[%v]` -> `%v`

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: `refreshedCluster :=`

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: would take this opportunity to change:
`[discovery] Synced Server and CertificateAuthorityData from the %s ConfigMap`

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig.go]: can be moved on the line above `if user == "" || len(....`

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig.go
Comment (neolit123) [inline cmd/kubeadm/app/util/kubeconfig/kubeconfig.go]: can be part of the above branch.
```
if config == nil || config.CurrentContext == "" || len(config.Contexts) == 0 || config.Contexts[config.CurrentContext] == nil {
...
```

**Review by neolit123 — COMMENTED** [review]
thanks @fabriziopandini 
added mostly some comments about stdout / test names.

File: cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go]: done, thanks

File: cmd/kubeadm/app/discovery/file/file.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/discovery/file/file.go]: kubeconfig is validated before this call, and this ensures only one authentication method is configured. Nevertheless, if in the future this `HasAuthenticationCredentials` it will be called somewhere else, I don't see problems because this function and `EnsureAuthenticationInfoAreEmbedded` have clearly separated scope

File: cmd/kubeadm/app/discovery/file/file.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/discovery/file/file.go]: filed https://github.com/kubernetes/kubeadm/issues/1693 to track this TODO

**Comment by fabriziopandini** [comment]
@neolit123 many thanks for the quick review! 
comments are addressed and answer to your questions are inline

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [reply cmd/kubeadm/app/discovery/file/file.go]: i guess in the case that i'm talking about the function `EnsureAuthenticationInfoAreEmbedded()` will fail with an `ioutil.ReadFile(authInfo.ClientCertificate)` error.

if the user has provided a kubeconfig with a CA and token it will pass the `HasAuthenticationCredentials` check, but then fail the `EnsureAuthenticationInfoAreEmbedded` check because the kubeconfig does not have client cert/key. so it seems to me this case is unsupported.

i think if `EnsureAuthenticationInfoAreEmbedded` fails here we should validate the token in the kubeconfig and only then exit the branch with an error.

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: `fileCluster :=`

**Comment by neolit123** [comment]
added another comment here: https://github.com/kubernetes/kubernetes/pull/80675#discussion_r308209562

the fix in https://github.com/kubernetes/kubernetes/pull/80066 seems related.
currently we don't load the CA from disk.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 
/hold 
^ for other reviewers, feel free to unhold when ready. 
+1 to backport.

File: cmd/kubeadm/app/discovery/file/file.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/discovery/file/file.go]: Clarified. EnsureAuthenticationInfoAreEmbedded does not fail in case of Kubeconfig file with token

**Comment by fabriziopandini** [comment]
@neolit123 
> the fix in #80066 seems related.

This fix focuses on CA, not on credentials, so they are orthogonal

**Review by hickeng — COMMENTED** [review]
I've read through the code change and did not spot any issues - I'm not sufficiently familiar with kubeadm code for that to mean much beyond a cursory check.

I think [this API doc](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/types.go#L360) needs updating as it's explicitly incorrect once this is merged.
> 	// TLSBootstrapToken is a token used for TLS bootstrapping.
	// If .BootstrapToken is set, this field is defaulted to .BootstrapToken.Token, but can be overridden.
	// If .File is set, this field **must be set** in case the KubeConfigFile does not contain any other authentication information

File: cmd/kubeadm/app/discovery/file/file.go
Comment (hickeng) [reply cmd/kubeadm/app/discovery/file/file.go]: `:=` construct looks more inline with existing inline declarations - is there a k8 style guide that makes a definitive statement?

**Comment by neolit123** [comment]
> I think this API doc needs updating as it's explicitly incorrect once this is merged.

>> // If .File is set, this field **must be set** in case the KubeConfigFile does not contain any other authentication information

the API comment still seems valid to me. see:
https://github.com/kubernetes/kubernetes/pull/80675/files#diff-fc46de7561d268ec1c3f3722181dd5aeR69

File: cmd/kubeadm/app/discovery/file/file.go
Comment (neolit123) [reply cmd/kubeadm/app/discovery/file/file.go]: we have this https://github.com/kubernetes/community/blob/master/contributors/guide/coding-conventions.md#code-conventions
which isn't very detailed, so as long as go{vet|lint|fmt} passes changes are tolerated.

`fileCluster :=` is preferred, so we left this one slide.

**Comment by hickeng** [comment]
> > I think this API doc needs updating as it's explicitly incorrect once this is merged.
> 
> > > // If .File is set, this field **must be set** in case the KubeConfigFile does not contain any other authentication information
> 
> the API comment still seems valid to me. see:
> https://github.com/kubernetes/kubernetes/pull/80675/files#diff-fc46de7561d268ec1c3f3722181dd5aeR69

If the bootstrap token is set then we return a new config that has the token instead of the client credentials per https://github.com/kubernetes/kubernetes/pull/80675/files#diff-fc46de7561d268ec1c3f3722181dd5aeR50

**Comment by neolit123** [comment]
@olivierlemasle there is an actual problem with this test job related to the build system.
please see:
https://kubernetes.slack.com/archives/C09QZ4DQB/p1564511430122100

we should retest once that is fixed.

**Comment by Boes-man** [comment]
Hello, is there a work around until 1.16? Thanks

**Comment by neolit123** [comment]
@Boes-man

> Hello, is there a work around until 1.16? Thanks

there are backports for 1.15 and 1.14.
for 1.15 not sure it will make it in .2, but maybe .3.

for 1.14 it should make it in the next patch release.

**Comment by Boes-man** [comment]
Awesome! Thanks

**Pull request merged**
